Automattic/sensei
# Integrate Sensei and MailPoet

Fixes #6425 

### Changes proposed in this Pull Request

* Provides MailPoet integration for Sensei Courses and Groups.
* Requires MailPoet plugin to be activated and set up with an API key from the MailPoet user dashboard.
* MailPoet automatically adds all WordPress users of the site as subscribers to a default 'WordPress' list. The logic in this integration therefore does not attempt to add new subscribers to MailPoet unless an exception is thrown.
* List synchronisation is done using the `SenseiScheduler` class and batch processing.

### Testing instructions

- Have groups and courses on your site.
- Add students to each group and course for testing MailPoet integration.
- Install MailPoet plugin.
- Go through the setup wizard, sign up for MailPoet and add API key to your plugin installation if you don't have an account.
- Visit MailPoet > Lists: 
  - Check to see that new lists are created for each course and group in your site.
- Visit MailPoet > Subscribers:
  - Make sure that all enrolled students in each course are subscribers to the course list on MailPoet. Check the same for each group.
  
 - Testing Enrolment Hooks
   - Enrol a student to a course
   - Check the MailPoet List page to see it added as a subscriber. MailPoet > Subscribers, select the course on the drop down filter and see the student added with status 'Unconfirmed'.
   - Remove the students' enrolment and check back in MailPoet to see status 'Unsubscribed'.
   - Repeat the same by adding and removing students from any group.
 

## New/Updated Hooks

* ~`sensei_integrate_mailpoet`~ `sensei_email_mailpoet_feature`: Determines whether to integrate MailPoet by adding lists for courses and groups.
  * Param: `boolean` `$enable` Enable this feature. Defaults to true.
  * Return: `boolean` Whether to enable feature.
* `sensei_email_mailpoet_sync_subscribers`: Schedule job to synchronise students in courses and groups with MailPoet subscribers list.
  * Params: None.


**UPDATE**
- Testing the Cron task
   - After one page load, use WP CLI to run the following command: `wp cron event list --fields=hook,next_run`
   - Check to see that the task `sensei_email_mailpoet_sync_subscribers` is scheduled to run.
   - Alternatively use a cron management plugin to see scheduled tasks, look for `sensei_email_mailpoet_sync_subscribers`.
   - Deactivate MailPoet plugin, check to see that the scheduled task is no longer there.
   - Activate MailPoet plugin, check that the task is back after one or two page loads.
   - Deactivate Sensei-LMS plugin, check to see that the scheduled task is no longer there.
